### PR TITLE
Add Item Detection Feature

### PR DESCRIPTION
**Pull request by brandonabela**

### CODE REVIEW

File: Logic/item_detection.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+from skimage.measure import compare_ssim as ssim
+
+from Config.eight_ball_lookup import EightBallLookup
+
+class ItemDetection:
+    '''Responsible for detecting game balls and holes'''
+
+    lookup = EightBallLookup()
+
+    def identify_parameters(self, for_holes, for_balls):
```
Comment (KeithGeorgeCiantar) [inline Logic/item_detection.py]: Add comment to clarify parameter type and use

File: Logic/item_detection.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        '''Responsible iterating through a number of images and saving the image results based on different parameter values'''
+
+        for ball_folder in os.listdir(self.lookup.training_path):
```
Comment (KeithGeorgeCiantar) [inline Logic/item_detection.py]: ball_folder might be misleading, consider, ball_image or ball_file, or maybe even just game_screenshot

File: Logic/item_detection.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+                cv2.imwrite(self.lookup.hole_training_path + str(param2) + '_' + str(param1) + '.jpg', rgb_image)
+
+    def find_ball_parameters(self, image_path, hole_positions):
+        '''Iterate through a range of values and saving each result in the goal of identify the best parameters for ball detection'''
+
+        min_x = min(hole_positions, key=lambda t: t[0])[0] + 10
+        min_y = min(hole_positions, key=lambda t: t[1])[1] + 10
+        max_x = max(hole_positions, key=lambda t: t[0])[0] - 10
+        max_y = max(hole_positions, key=lambda t: t[1])[1] - 10
```
Comment (KeithGeorgeCiantar) [inline Logic/item_detection.py]: Add comment to signify that these values correspond to the corners

File: Logic/item_detection.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+                circles = cv2.HoughCircles(gray_image, cv2.HOUGH_GRADIENT, 1, 10, param1=param1, param2=param2, minRadius=9, maxRadius=13)
+
+                if circles is not None:
+                    circles = np.round(circles[0, :]).astype("int")
+
+                    for (x_position, y_position, radius) in circles:
+                        radius = 13 # assuming ball is always 13 pixels
+
+                        if min_x <= x_position and x_position <= max_x and min_y <= y_position and y_position < max_y:
```
Comment (KeithGeorgeCiantar) [inline Logic/item_detection.py]: Add <= for y_position and max_y as well

File: Logic/item_detection.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+
+    def detect_balls(self, frame, hole_positions):
+        '''Responsible for returning an array of ball positions'''
+
+        detected_circles = []
+
+        min_x = min(hole_positions, key=lambda t: t[0])[0] + 10
+        min_y = min(hole_positions, key=lambda t: t[1])[1] + 10
+        max_x = max(hole_positions, key=lambda t: t[0])[0] - 10
+        max_y = max(hole_positions, key=lambda t: t[1])[1] - 10
```
Comment (KeithGeorgeCiantar) [inline Logic/item_detection.py]: Refactor, make a function that gives the 4 corners in a list and call it

File: Logic/item_detection.py [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+        circles = cv2.HoughCircles(gray_image, cv2.HOUGH_GRADIENT, 1, 9, param1=300, param2=15, minRadius=9, maxRadius=13)
+
+        if circles is not None:
+            circles = np.round(circles[0, :]).astype("int")
+
+            for (x_position, y_position, radius) in circles:
+                radius = 13 # assuming ball is always 13 pixels
+
+                if min_x <= x_position and x_position <= max_x and min_y <= y_position and y_position < max_y:
```
Comment (KeithGeorgeCiantar) [inline Logic/item_detection.py]: Similar to before <= for all

File: Logic/item_detection.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+                print('Current Timestamp: ' + str(time_hours) + ':' + str(time_minutes) + ':' + str(time_seconds))
+
+            if not holes: # Only detect the balls once
+                holes = self.detect_holes(frame)
+
+                min_x = min(holes, key=lambda t: t[0])[0]
+                min_y = min(holes, key=lambda t: t[1])[1]
+                max_x = max(holes, key=lambda t: t[0])[0]
+                max_y = max(holes, key=lambda t: t[1])[1]
```
Comment (KeithGeorgeCiantar) [inline Logic/item_detection.py]: See refactoring note

**Review by KeithGeorgeCiantar — CHANGES_REQUESTED** [review]
I have attached some comments, you should be able to see them

File: Logic/item_detection.py (hunk 1 above)
Comment (brandonabela) [reply Logic/item_detection.py]: Done

File: Logic/item_detection.py (hunk 2 above)
Comment (brandonabela) [reply Logic/item_detection.py]: Done

File: Logic/item_detection.py (hunk 3 above)
Comment (brandonabela) [reply Logic/item_detection.py]: Done

File: Logic/item_detection.py (hunk 4 above)
Comment (brandonabela) [reply Logic/item_detection.py]: Done

File: Logic/item_detection.py (hunk 5 above)
Comment (brandonabela) [reply Logic/item_detection.py]: Done

File: Logic/item_detection.py (hunk 6 above)
Comment (brandonabela) [reply Logic/item_detection.py]: Done

File: Logic/item_detection.py (hunk 7 above)
Comment (brandonabela) [reply Logic/item_detection.py]: Done

**Review by KeithGeorgeCiantar — APPROVED** [review]
Code seems much neater, good job on refactoring, cleanup and reformatting.

**Pull request merged**